### PR TITLE
Fix assignments to RealmList properties on managed objects

### LINKED ISSUE #718: Setting a new list fails to modify list on managed object
It looks like the underlying issue is that lists just aren't updated on managed objects, no matter if the elements of the list are managed or not.
```

    @Test
    fun updateListOnManagedObjectTest() {
        realm.writeBlocking {
            val thingHolder = ThingHolder().apply { listOfThings = realmListOf() }
            copyToRealm(thingHolder)
        }

        realm.writeBlocking {

            val thing1 = Thing().apply { id = "thing1" }
            val thing1Managed = copyToRealm(thing1)
            val thingHolder = realm.query<ThingHolder>()
                .first()
                .find()!!
                .let { findLatest(it)!! }

            thingHolder.apply {
                name = "updatedName"
                listOfThings = realmListOf(thing1Managed)
            }

            copyToRealm(thingHolder)
        }

        val thingHolder = realm.query<ThingHolder>()
            .first()
            .find()!!

        assertEquals(thingHolder.name, "updatedName") // Passes
        assertNotNull(thingHolder.listOfThings.firstOrNull()) // Fails
    }
```

### PR DESCRIPTION
**Pull request by rorbech**
This adds support for assigning and importing unmanaged objects by assigning `RealmList` property on an already managed object.

Closes #718

### CODE REVIEW

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt [hunk 1]
```diff
@@ -329,7 +329,10 @@ internal object RealmObjectHelper {
     }
 
     @Suppress("UNUSED_PARAMETER")
```
Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:331]: Is this annotation still relevant?

Comment (clementetb) [inline packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:331]: Is this annotation still relevant?

File: test/base/src/androidTest/kotlin/io/realm/test/shared/RealmListTests.kt [hunk 2]
```diff
@@ -819,6 +858,31 @@ internal abstract class ManagedListTester<T>(
...
 
+    override fun reassign() {
+        val dataSet = typeSafetyManager.getInitialDataSet()
+        val reassignedDataSet = listOf(dataSet[1])
+
+        val assertions = { list: RealmList<T> ->
+            assertEquals(1, list.size)
+            // We cannot assert equality on RealmObject lists as the object isn't equals to the
+            // unmanaged object from before the assignment
+            if (list[0] !is RealmObject) {
```
Comment (cmelchior) [inline test/base/src/androidTest/kotlin/io/realm/test/shared/RealmListTests.kt:869]: Would it be possible to check some property on the object instead, just as a smoke test? Right now RealmObjects are kinda slipping through this test.

**Review by cmelchior — APPROVED** [review]
Nice....I'm not 100% sure `reassign` should be in the ListTests, but we don't have another more obvious place right now, and probably speak to a broader discussion.

**Review by clementetb — APPROVED** [review]
LGTM, happy if CI is happy

File: test/base/src/androidTest/kotlin/io/realm/test/shared/RealmListTests.kt (hunk 2 above)
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/RealmListTests.kt:869]: We already assert that the list is actually updated since the initial lists have two elements (guaranteed by accessing index 1) and the resulting list only having one (`list.size == 1` from above assert). But I can probably find something to assert.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt (hunk 1 above)
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:331]: Yes. It is only invoked from compiler plugin generated code. I will add a comment.

**Comment by rorbech** [comment]
> Nice....I'm not 100% sure `reassign` should be in the ListTests, but we don't have another more obvious place right now, and probably speak to a broader discussion.

Yeah. I though of it, but didn't want to put up the whole _exhaustive type_ infrastructure somewhere else. I guess we should get the infrastructure migrated to use the schema API and reassess where to put these.

**Comment by rorbech** [comment]
The existing tests showed some issues with assignments from already managed list and especially self assignments resulting in empty lists. Added some `RealmInterop.realm_equals` checks and some additional test of specific assignment scenarios. 

Haven't touched `DynamicMutableRealmObject.set` so that still doesn't support updating lists. Requires a bit more effort to split paths based on type (or name `Value/Object/List` variants?) along with all the tests, so I would suggest filing a separate issue on that to avoid blocking this fix. All list operations are also still doable on the list retrieved by `getList` so we have an immediate workaround for that.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt [hunk 3]
```diff
@@ -328,8 +328,14 @@ internal object RealmObjectHelper {
         }
     }
 
-    @Suppress("UNUSED_PARAMETER")
-    internal fun setList(obj: RealmObjectInternal, col: String, list: RealmList<Any?>) {
-        TODO()
+    @Suppress("unused") // Called from generated code
+    inline fun <reified T : Any> setList(obj: RealmObjectInternal, col: String, list: RealmList<Any?>) {
+        val existingList = getList<T>(obj, col)
+        if (list !is ManagedRealmList || !RealmInterop.realm_equals(existingList.nativePointer, list.nativePointer)) {
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:334]: I don't think this check is correct..imagine this usecase, which I would expect to work:

```
realm.write {
  val obj1 = copyToRealm(Person("Mom"))
  val obj2 = copyToRealm(Person("Data))
  obj1.children.add(Person("Mia"))
  obj2.children = obj1.children
}
```

Looks like we are also missing a test for this.

File: test/base/src/androidTest/kotlin/io/realm/test/shared/SampleTests.kt [hunk 4]
```diff
@@ -257,4 +259,47 @@ class SampleTests {
...
+    // @Test
+    // fun list_assign_allTypes() {}
+
+    @Test
+    fun list_assign_unmanaged() {
+        realm.writeBlocking {
+            val objectList: List<Sample> = (0..9).map { Sample().apply { intField = it } }
+            val sample = copyToRealm(Sample()).apply { objectListField = objectList.toRealmList() }
+        }
+        assertEquals(11, realm.query<Sample>().count().find())
```
Comment (cmelchior) [inline test/base/src/androidTest/kotlin/io/realm/test/shared/SampleTests.kt:274]: Ahh, good old inclusive `..` in both ends 🙄

**Review by cmelchior — CHANGES_REQUESTED** [review]
Nice...but I think I identified a corner case that isn't supported. But apart from that everything looks good.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt (hunk 3 above)
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:334]: I am pretty sure that it is right, and that it is indirectly covered by the test in https://github.com/realm/realm-kotlin/pull/719/files#diff-a9009c4bd0a6af3222c32304364a65fe380a3f5ecdd1e84db0f5356ada531193R278 ... but lets see when I test it explicitly 🙈

File: test/base/src/androidTest/kotlin/io/realm/test/shared/SampleTests.kt (hunk 4 above)
Comment (rorbech) [reply test/base/src/androidTest/kotlin/io/realm/test/shared/SampleTests.kt:274]: Yeah. _All inclusive_ seems to be the new standard for next generations 👴

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt (hunk 3 above)
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:334]: Reinspecting the test in https://github.com/realm/realm-kotlin/pull/719/files#diff-a9009c4bd0a6af3222c32304364a65fe380a3f5ecdd1e84db0f5356ada531193R278 and it already seem to update a managed object assigning a managed list from object to another list. Isn't it the linked test exactly what you are looking for?

Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:334]: Not 100%, there you assign a managed list to an unmanaged object here https://github.com/realm/realm-kotlin/pull/719/files#diff-a9009c4bd0a6af3222c32304364a65fe380a3f5ecdd1e84db0f5356ada531193R287... I was looking for a test that assigned a managed list to an managed object

Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:334]: But the object is managed. The `copyToRealm(Sample())` returns a managed object and the assignment is inside the `apply` block on that.

Comment (cmelchior) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:334]: Ah yes 🙈 ...I missed a brace there ... Yes, that means the case I was thinking about is tested. Sorry for the confusion.

**Comment by rorbech** [comment]
> Nice...but I think I identified a corner case that isn't supported. But apart from that everything looks good.

@cmelchior As per https://github.com/realm/realm-kotlin/pull/719#discussion_r824728455 I think this should be good to go?

**Review by cmelchior — APPROVED** [review]
:shipit:
